aws/resource_aws_waf_rule.go

```diff
@@ -199,11 +209,24 @@ func resourceAwsWafRuleDelete(d *schema.ResourceData, meta interface{}) error {
 			ChangeToken: token,
 			RuleId:      aws.String(d.Id()),
 		}
-		log.Printf("[INFO] Deleting WAF Rule")
-		return conn.DeleteRule(req)
+
+		output, err := conn.DeleteRule(req)
+
+		// Deleting a WAF Rule after being removed from a WAF WebACL
+		// can return a WAFReferencedItemException when attempted in quick succession;
+		// thus, we catch the error here and re-attempt
+		if tfawserr.ErrCodeEquals(err, waf.ErrCodeReferencedItemException) {
+			return output, nil
+		}
```

Comment:
Can you explain how this error is being retried? The `RetryWithToken` helper currently only handles the `waf.ErrCodeStaleDataException` error and this logic is swallowing `waf.ErrCodeReferencedItemException`. I'm guessing this might be passing `TestAccAWSWafWebAcl_Rules` because that test uses `CheckDestroy: testAccCheckAWSWafWebAclDestroy` and does not verify rule deletion.

This logic would need another layer of `resource.Retry()` to retry successfully on `waf.ErrCodeReferencedItemException` and it might be worth creating a covering `TestAccAWSWafRule_` test for this behavior so it can use `CheckDestroy: testAccCheckAWSWafRuleDestroy,`. 👍

Reply:
ahh cannot be explained ..totally in the wrong place there! but I forgot i've wrestled with the `TestAccAWSWafWebACL_Rules` test quite some time ago and from what i've tried it's a case where if the test config steps were applied outside of the test env (i've most recently used the latest version of the provider w/terraform 0.14), the resource modifications occur without any error.

Reply:
Personally, I'd love to get the rest of these PR changes in just so those refactorings are out of the way for the future. Up to you if you'd like to revert this small little bit so we can get those in or if you would prefer working on this section some more here. Feel free to grab me directly to chat more!

Reply:
yep let me keep the refactorings in and remove the attempted test fix 👍

Reply:
...or may be not 😅 ..i _think_ 62adf9a should correctly use the retry and the `TestAccWafRule_webACL` added for additional checking